pkg/vfs/metadata.go

```diff
@@ -181,3 +186,82 @@ func (e *ExifExtractor) Result() Metadata {
 	}
 	return m
 }
+
+// AudioExtractor is used to extract album/artist/etc. from audio
+type AudioExtractor struct {
+	w  *io.PipeWriter
+	r  *io.PipeReader
+	ch chan interface{}
+}
+
+// NewAudioExtractor returns an extractor for audio
+func NewAudioExtractor() *AudioExtractor {
+	e := &AudioExtractor{}
+	e.r, e.w = io.Pipe()
+	e.ch = make(chan interface{})
+	go e.Start()
+	return e
+}
+
+// Start is used in a goroutine to start the metadata extraction
+func (e *AudioExtractor) Start() {
+	buf, err := ioutil.ReadAll(e.r)
+	if err != nil {
+		e.ch <- err
+		return
+	}
+	tags, err := tag.ReadFrom(bytes.NewReader(buf))
+	e.r.Close()
+	if err != nil {
+		e.ch <- err
+	} else {
+		e.ch <- tags
+	}
+}
+
+// Write is called to push some bytes to the extractor
+func (e *AudioExtractor) Write(p []byte) (n int, err error) {
+	return e.w.Write(p)
+}
+
+// Close is called when all the bytes has been pushed, to finalize the extraction
+func (e *AudioExtractor) Close() error {
+	return e.w.Close()
+}
+
+// Abort is called when the extractor can be discarded
+func (e *AudioExtractor) Abort(err error) {
+	e.w.CloseWithError(err)
+	<-e.ch
+}
+
+// Result is called to get the extracted metadata
+func (e *AudioExtractor) Result() Metadata {
+	m := NewMetadata()
+	tags := <-e.ch
+	switch tags := tags.(type) {
+	case tag.Metadata:
```

Comment:
Well, it would work too. The switch case on type may be easier to extend if we had to use several libs for extracting audio metadata (e.g. one for id3 and another for flac).